### PR TITLE
fix: suggested smallcaps improvement

### PR DESCRIPTION
**Pull request by erights**
Based on a suggestion by @mhofman at https://github.com/endojs/endo/pull/1282#discussion_r968921875

### CODE REVIEW

File: packages/marshal/src/encodeToSmallcaps.js [hunk 1]
```diff
@@ -32,37 +28,48 @@ import {
...
+ *  * `+` - non-negative bigint
+ *  * `-` - negative bigint
+ *  * `#` - manifest constant
+ *  * `%` - symbol
+ *  * `$` - remotable
+ *  * `&` - promise
+ *
+ * All other special characters (`!"()*,`) are reserved for future use.
+ *
+ * The menifest constants that smallcaps currently uses for values:
```
Comment (turadg) [inline packages/marshal/src/encodeToSmallcaps.js:49]: manifest

**Comment by erights** [comment]
This one is just better, so I'm merging it into https://github.com/endojs/endo/pull/1282 as completion of https://github.com/endojs/endo/pull/1282#discussion_r968921875 review comment.

File: packages/marshal/src/encodeToSmallcaps.js [hunk 2]
```diff
@@ -292,177 +296,138 @@ export const makeDecodeFromSmallcaps = ({
...
+            const result = decodePromiseFromSmallcaps(encoding);
+            if (passStyleOf(result) !== 'promise') {
+              assert.fail(
+                X`internal: decodePromiseFromSmallcaps option must return a promise: ${result}`,
+              );
             }
+            return result;
+          }
+          default: {
+            return encoding;
```
Comment (mhofman) [inline packages/marshal/src/encodeToSmallcaps.js:363]: Should we enforce that the string doesn't start with one of the other currently unused hilberted chars? Maybe we can add `case` statements for those ?

File: packages/marshal/src/encodeToSmallcaps.js (hunk 1 above)
Comment (erights) [reply packages/marshal/src/encodeToSmallcaps.js:49]: Done

File: packages/marshal/src/encodeToSmallcaps.js [hunk 3]
```diff
@@ -32,37 +28,48 @@ import {
...
- * @returns {encoded is SmallcapsEncodingUnion}
- */
-const hasSClass = encoded => hasOwnPropertyOf(encoded, SCLASS);
+const startsSpecial = encodedStr => {
+  if (encodedStr === '') {
+    return false;
+  }
+  const c = encodedStr[0];
+  // eslint-disable-next-line yoda
+  return '!' <= c && c <= '-';
```
[the remark this answers is not in the document]
Comment (erights) [reply packages/marshal/src/encodeToSmallcaps.js:71]: Do we have any info on whether that's true on XS?

File: packages/marshal/src/encodeToSmallcaps.js [hunk 4]
```diff
@@ -130,18 +137,28 @@ export const makeEncodeToSmallcaps = ({
     switch (passStyle) {
       case 'null':
       case 'boolean': {
+        // pass through to JSON
         return passable;
       }
       case 'string': {
-        if (passable !== '' && SpecialChars.includes(passable[0])) {
+        if (startsSpecial(passable[0])) {
```
[the remark this answers is not in the document]
Comment (erights) [reply packages/marshal/src/encodeToSmallcaps.js:144]: Done

File: packages/marshal/src/encodeToSmallcaps.js [hunk 5]
```diff
@@ -159,47 +176,31 @@ export const makeEncodeToSmallcaps = ({
...
         );
       }
       case 'copyArray': {
         return passable.map(encodeToSmallcapsRecur);
       }
       case 'tagged': {
         return {
-          [SCLASS]: 'tagged',
-          tag: getTag(passable),
+          '#tagged': getTag(passable),
```
[the remark this answers is not in the document]
Comment (erights) [reply packages/marshal/src/encodeToSmallcaps.js:203]: Only blind spot. Good idea. Done.

File: packages/marshal/src/encodeToSmallcaps.js [hunk 6]
```diff
@@ -292,177 +296,138 @@ export const makeDecodeFromSmallcaps = ({
...
-            case '#-Infinity': {
-              return -Infinity;
+          }
+          case '+':
+          case '-': {
+            const last = encoding.length - 1;
+            assert(
+              encoding[last] === 'n',
+              X`Encoded bigint must start with "+" or "-" and end with "n": ${encoding}`,
+            );
```
[the remark this answers is not in the document]
Comment (erights) [reply packages/marshal/src/encodeToSmallcaps.js:341]: Already done

File: packages/marshal/src/encodeToSmallcaps.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (mhofman) [reply packages/marshal/src/encodeToSmallcaps.js:71]: As a hunch I would expect it to be true in most engines, as allocating a string and comparing strings feel more expensive than numbers.

File: packages/marshal/src/encodeToSmallcaps.js (hunk 2 above)
Comment (erights) [reply packages/marshal/src/encodeToSmallcaps.js:363]: Done

File: packages/marshal/src/encodeToSmallcaps.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (erights) [reply packages/marshal/src/encodeToSmallcaps.js:71]: This one makes the code enough uglier that I'm not gonna do it until and unless we have measurements that say it is significant for us on XS. Resolving.

[the remark this answers is not in the document]
Comment (mhofman) [reply packages/marshal/src/encodeToSmallcaps.js:71]: What I've done in other places is to define consts at the top level and use that for comparisons.

```js
const EXCLAMATION = '!'.charCodeAt(0);
const DASH = '-'.charCodeAt(0);
```

Then:
```js
  const code = encodedStr.charCodeAt(0);
  // eslint-disable-next-line yoda
  return EXCLAMATION <= code && code <= DASH;
```

**Pull request merged**